### ISSUE TITLE
[top/rom] Allow software to delay device debug access 

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#13634]
We currently enable debug access purely based on life cycle.
Specifically, when we are in TEST / DEV / RMA states, debug access is automatically enabled. 

However, during production states (or any state where the device might be widely distributed, such as DEV state potentially), we can further delay debug enablement to give software a chance to fully wipe out any relevant content. 

This is beneficial because a device reset does not necessarily wipe non-flop contents like memory, and those may hold important secrets that we do not want to expose to debug access.  This is especially true for RMA transitions, where post transition a reset is required.

The trade-off to such a feature is that ROM debugging would be hampered. If JTAG for example is not enabled until ROM executes, then ROM by definition cannot be debugged. 

This is a security / usability trade off that needs to be made.  For more mature projects, it will make more sense to lean towards security instead of usability.

**Comment by alphan** [comment]
> This is especially true for RMA transitions, where post transition a reset is required.

IIRC we erase the owner flash when transitioning to RMA, can we erase any memories that we are worried about during that process? Are we concerned about the additional hardware this might require?

**Comment by tjaychen** [comment]
yes that's something we added to `otbn`. 
So we filed this issue as a future enhancement to see if there are even more secure approaches.

**Comment by mundaym** [comment]
This issue was briefly discussed in the software working group (2022-08-02). The consensus was to wait for suggestions from the hardware teams before discussing further.

**Comment by andreaskurth** [comment]
Triaged for `rv_dm`. Assigning to M2.5 with https://github.com/lowRISC/opentitan/labels/Priority%3AP1 because I think we should make sure current HW and ROM are suitable for production.  If yes, defer back to FutureRelease.

**Comment by gdessouky** [comment]
Is the concern here only non-volatile flash memory and retention SRAM? Or are any other memories the issue, e.g., OTBN SRAM? Can we reuse the same mechanism that erases owner flash content before RMA transition also for DEV and TEST transitions? The next question that comes to mind is whether we want to even do that, might it impact functionality if we erase flash content at DEV and TEST transitions where owner might not be changing?

Also, all memories are scrambled. This also offers some level of confidentiality, though not cryptographically strong AFAIK. Unless we can identify a tangible security implication, I would go for usability and accessible ROM debug access with this one. 

@andreaskurth I second your suggestion that we discuss this in the Security WG and hear some more thoughts to decide.

**Comment by moidx** [comment]
I prefer not to block our ability to debug the ROM until we have tested on silicon. The following issues capture previous discussions: 

* https://github.com/lowRISC/opentitan/issues/15222
* https://github.com/lowRISC/opentitan/issues/15613

**Comment by andreaskurth** [comment]
As discussed in the Security WG on 2023-03-16, we don't see a reason to delay debug access in software (in particular enable it only after a part of ROM has executed).

Debug access is [only enabled in the TEST_UNLOCKED, DEV, or RMA states](https://opentitan.org/book/doc/security/specs/device_life_cycle/index.html).

In TEST_UNLOCKED and DEV, the device does not contain secrets that would be protected by delaying debug access post ROM execution (those states cannot be reached from PROD).

RMA entry is protected by a token. Upon RMA entry, secret partitions of flash get wiped, key manager derivation changes, and OTBN changes the scramble keys for its memories. The SRAM controller resets to its default scrambling key. If that key is different from an ephemeral scrambling key provided by OTP prior to reset, SRAM content is rescrambled as well. There are thus multiple protection mechanisms against leaking confidential data through debug access in RMA.

Delaying debug access in software would prevent debug access to the core prior to debug enablement in software. This reduces observability in debug environments and would also influence our manufacturing programs.

For all these reasons, we have decided to close this issue. An option worth considering for future releases could be to distinguish TEST from DEV w.r.t. debug enablement. If we see a need for this, we'll open a new issue.
